### PR TITLE
[codex] Add Arthexis end and cleanup turn skills

### LINKED ISSUE #7414: Add end-step and cleanup-step to Arthexis turn sequence
## Intent
Add explicit end-step and cleanup-step phases to the Arthexis turn sequence.

## Scope
- Turn skills and sequencing documentation
- Any scripts that execute turn/next-turn behavior
- Process tracking for work started during a turn

## Acceptance Criteria
- End-step runs after all main phases.
- Effects declared to happen at end of turn trigger during end-step before cleanup.
- Cleanup-step runs last.
- Cleanup waits for turn-owned processes to finish cleanly, then terminates lingering turn-owned processes after up to 10 minutes.
- Explicit turn/next-turn execution includes untap-step, upkeep-step, draw-step, main-phase(s), end-step, and cleanup-step in order.

## Constraints
- Do not affect wlan1 or network connectivity.
- Do not implement more than one TODO per main phase.
- Keep /home/arthe/TODO.md as the local queue until the suite TODO app exists.

### PR DESCRIPTION
**Pull request by arthexis**
## Summary
- adds an end-step skill that triggers declared end-of-turn effects before cleanup
- adds a cleanup-step skill with turn-owned PID registration, up-to-10-minute wait, and bounded termination of only registered turn-owned processes
- archives completed turn state under the local Arthexis turn state directory

## Issue
Closes #7414

## Verification
- python3 -c "from pathlib import Path; path=Path("skills/arthexis-cleanup-step/scripts/turn_boundary.py"); compile(path.read_text(encoding="utf-8"), str(path), "exec"); print("syntax: ok")"
- python3 skills/arthexis-cleanup-step/scripts/turn_boundary.py status
- git diff --check

## Local device validation
- Installed matching local skills under /home/arthe/.codex/skills.
- Ran a synthetic turn with one declared end effect and cleanup --timeout 0; the effect triggered and cleanup archived the turn without touching processes.

### CODE REVIEW

**Comment by arthexis** [comment]
Worker update for the SonarCloud security gate on PR #7421.

SonarCloud details checked via its API showed two open hotspots in `skills/arthexis-cleanup-step/scripts/turn_boundary.py`:
- `python:S4828` at the SIGTERM send path: "Make sure that sending signals is safe here."
- `python:S4828` at the SIGKILL send path: "Make sure that sending signals is safe here."

Patch pushed: `f13b0cee4` (`Harden turn cleanup process signaling`). The cleanup script now carries PID identity snapshots through the wait, revalidates PID/start-ticks/UID immediately before each signal, refuses PID 1 and the current process, and keeps the existing bounded behavior of signaling only registered turn-owned processes and their descendants. I also added the narrow stale-state guard from the high-priority Gemini cleanup race comment so a concurrent cleanup cannot archive over a completed or changed active turn.

Verification run inside Docker container `ridge-pr7421` only:
- `python3 -c "from pathlib import Path; path=Path(\"skills/arthexis-cleanup-step/scripts/turn_boundary.py\"); compile(path.read_text(encoding=\"utf-8\"), str(path), \"exec\"); print(\"syntax: ok\")"`
- `python3 -m py_compile skills/arthexis-cleanup-step/scripts/turn_boundary.py`
- `python3 skills/arthexis-cleanup-step/scripts/turn_boundary.py status`
- synthetic turn with declared end effect and `cleanup-step --timeout 0`, confirming archive/no-active-turn status
- registered container-local `sleep` process followed by `cleanup-step --timeout 0`, confirming SIGTERM was sent to the registered process and it exited
- `git diff --check`

Residual risk: I cannot locally rerun SonarCloud itself; the pushed commit should trigger the external SonarCloud check to confirm the hotspot count is cleared.

**Comment by arthexis** [comment]
Follow-up after the first push: SonarCloud re-reported the same `python:S4828` hotspots on the moved `os.kill` calls, so I pushed `6841227e8` (`Use pidfd for turn cleanup signaling`).

The signal path now uses Linux `pidfd_open` + `signal.pidfd_send_signal` after revalidating PID/start-ticks/UID, then revalidates again after opening the pidfd before sending. This removes numeric-PID `os.kill` signaling from the cleanup path while preserving the intended bounded cleanup behavior for registered turn-owned processes.

I reran the same container verification after this follow-up:
- syntax compile command for `skills/arthexis-cleanup-step/scripts/turn_boundary.py`
- `python3 -m py_compile skills/arthexis-cleanup-step/scripts/turn_boundary.py`
- cleanup `status` command
- synthetic end/cleanup archive run
- registered container-local `sleep` cleanup run
- `git diff --check`

Residual risk remains the external SonarCloud rerun; the fresh commit should give Sonar no `os.kill` calls to report for this hotspot rule.

**Comment by arthexis** [comment]
Follow-up for the remaining #7421 blockers.

Pushed `d0bc5aab8` (`Make turn cleanup skill paths portable`) on top of the earlier signal hardening commits.

What changed:
- Addressed `PRRT_kwDOPUOpW859k-Fx` and `PRRT_kwDOPUOpW859k-F0` by replacing hardcoded `/home/arthe` examples/state paths in the cleanup/end skill docs with `~` paths.
- Addressed `PRRT_kwDOPUOpW859k-Fz` by changing `STATE_DIR` to `Path.home() / ".local" / "state" / "arthexis-turn"`.
- Re-checked the race-condition thread `PRRT_kwDOPUOpW859k-Fu`: latest code no longer falls back to stale pre-wait state after reacquiring the lock. It aborts with `status: turn-completed-or-changed` if another cleanup completed or replaced the active turn, so it should be ready for overseer resolution once checks are clean.
- Re-investigated SonarCloud after `6841227e8`: the previous API result still showed stale hotspots at the old `os.kill` lines, but after `d0bc5aab8` SonarCloud hotspot API returns `total: 0` for PR #7421.

Verification run inside Docker container `ridge-pr7421` only:
- syntax compile command for `skills/arthexis-cleanup-step/scripts/turn_boundary.py`
- `python3 -m py_compile skills/arthexis-cleanup-step/scripts/turn_boundary.py`
- `python3 skills/arthexis-cleanup-step/scripts/turn_boundary.py status`
- synthetic turn with declared end effect and `cleanup-step --timeout 0`, confirming archive/no-active-turn status under the executing user home
- registered container-local `sleep` followed by `cleanup-step --timeout 0`, confirming bounded cleanup terminates the registered process
- `grep -RInE "/home/arthe|os\\.kill" skills/arthexis-cleanup-step skills/arthexis-end-step || true` returned no matches
- `git diff --check`

Residual risk: GitHub checks for the latest push are still running at comment time; SonarCloud hotspot API currently reports no open hotspots for the PR.

**Comment by arthexis** [comment]
Additional Sonar follow-up pushed: `1129f66bb` (`Constrain turn state archive paths`).

Investigation after `d0bc5aab8` showed the two signal hotspots were cleared (`security_hotspots` API returned `total: 0`), but the quality gate still failed because Sonar reported a separate blocker vulnerability:
- `pythonsecurity:S2083` in `skills/arthexis-cleanup-step/scripts/turn_boundary.py`, message: "Change this code to not construct the path from user-controlled data." The taint flow pointed from active turn JSON into archive path construction/write.

Fix:
- added state-path validation so JSON reads/writes are constrained under `Path.home() / ".local" / "state" / "arthexis-turn"`
- validates turn IDs with a narrow filename-safe pattern before accepting `--turn-id` or archiving a turn state
- rejects path-like turn IDs such as `../bad` before any state is written

Verification run inside Docker container `ridge-pr7421` only:
- syntax compile command for `skills/arthexis-cleanup-step/scripts/turn_boundary.py`
- `python3 -m py_compile skills/arthexis-cleanup-step/scripts/turn_boundary.py`
- cleanup `status` command
- synthetic declared-effect/end-step/cleanup archive flow
- invalid turn id probe: `start-turn --turn-id ../bad --force` returns status 2
- registered container-local `sleep` cleanup run
- no `/home/arthe` or `os.kill` matches in owned skill dirs
- `git diff --check`

I did not resolve review threads, mark ready, merge, or change labels. External checks are expected to rerun on this latest head.

**Comment by arthexis** [comment]
Follow-up pushed: `2e926905a` (`Use fixed turn state temp path`).

Sonar reanalysis on `1129f66bb` still reported `pythonsecurity:S2083` at the temp JSON write line. I kept the prior state-root and turn-id validation, and also removed the destination-derived temporary filename. `write_json()` now writes through a fixed temp path inside the validated turn state directory before replacing the validated target path, so neither the temp file path nor archive filename can be influenced by JSON-loaded turn state.

Verification rerun inside Docker container `ridge-pr7421` only:
- syntax compile command for `skills/arthexis-cleanup-step/scripts/turn_boundary.py`
- `python3 -m py_compile skills/arthexis-cleanup-step/scripts/turn_boundary.py`
- cleanup `status` command
- synthetic declared-effect/end-step/cleanup archive flow
- invalid turn id probe rejects `../bad` with status 2
- registered container-local `sleep` cleanup run
- no `/home/arthe` or `os.kill` matches in owned skill dirs
- `git diff --check`

I did not resolve threads, mark ready, merge, or change labels.

**Comment by arthexis** [comment]
Follow-up pushed: `9959c2419` (`Avoid path-derived temp writes`).

Sonar reanalysis on `2e926905a` still tracked `pythonsecurity:S2083` to the temp write sink. I removed the `Path.write_text()` temp-file path from this flow entirely. `write_json()` now:
- validates the target path under the turn state root
- limits writes to the state root or archive directory
- writes through `os.open(..., dir_fd=...)` using the constant filename `.write-json.tmp`
- atomically replaces the validated target filename in the already-open validated directory

Verification rerun inside Docker container `ridge-pr7421` only:
- syntax compile command for `skills/arthexis-cleanup-step/scripts/turn_boundary.py`
- `python3 -m py_compile skills/arthexis-cleanup-step/scripts/turn_boundary.py`
- cleanup `status` command
- synthetic declared-effect/end-step/cleanup archive flow
- invalid turn id probe rejects `../bad` with status 2
- registered container-local `sleep` cleanup run
- no `/home/arthe` or `os.kill` matches in owned skill dirs
- `git diff --check`

I did not resolve threads, mark ready, merge, or change labels. Waiting on external Sonar reanalysis for latest head `9959c2419`.

File: skills/arthexis-cleanup-step/scripts/turn_boundary.py [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+        state["cleanup"] = {
+            "timeout_seconds": timeout_seconds,
+            "live_before": live_before,
+            "lingering_after_wait": sorted(lingering),
+            **termination,
+        }
+        archive_state(state)
+        if ACTIVE_STATE.exists():
+            ACTIVE_STATE.unlink()
+        append_event({"event": "cleanup-complete", "turn_id": state["turn_id"], "cleanup": state["cleanup"]})
```
[the remark this answers is not in the document]
Comment (arthexis) [reply skills/arthexis-cleanup-step/scripts/turn_boundary.py]: Addressed across the latest cleanup commits, ending at 9959c2419: cleanup no longer falls back to stale pre-wait state after reacquiring the lock; if another cleanup completed or replaced the active turn, it aborts instead of archiving stale data. Worker verification covered synthetic cleanup/archive flows.

File: skills/arthexis-cleanup-step/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+- It must not kill unrelated host processes, stop services, edit packages, modify repositories, write application databases, change network policy, reboot, or affect `wlan1`.
+- If a process was not registered as turn-owned, report it as outside cleanup authority instead of terminating it.
+
+## Workflow
+
+1. Re-read `/home/arthe/AGENTS.md` and `/home/arthe/workgroup.txt` when coordination may matter.
+2. Run cleanup with the default 10-minute wait:
+
+```bash
+python3 /home/arthe/.codex/skills/arthexis-cleanup-step/scripts/turn_boundary.py cleanup-step --timeout 600
```
[the remark this answers is not in the document]
Comment (arthexis) [reply skills/arthexis-cleanup-step/SKILL.md]: Addressed in the portability follow-up: cleanup skill examples now use home-relative paths instead of hardcoded `/home/arthe` paths. Worker verification also checked no `/home/arthe` matches remain in the owned skill dirs.

File: skills/arthexis-cleanup-step/scripts/turn_boundary.py [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+import signal
+import sys
+import time
+import uuid
+from contextlib import contextmanager
+from pathlib import Path
+from typing import Any
+
+
+STATE_DIR = Path("/home/arthe/.local/state/arthexis-turn")
```
[the remark this answers is not in the document]
Comment (arthexis) [reply skills/arthexis-cleanup-step/scripts/turn_boundary.py]: Addressed in the portability follow-up: `STATE_DIR` is now built from `Path.home()` and writes are constrained under the user turn-state root. Worker verification covered status, archive, invalid turn-id, and cleanup flows.

File: skills/arthexis-end-step/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+- Not allowed: package changes, service changes, process termination, network/firewall edits, database writes, credential changes, branch changes, GitHub mutations, destructive cleanup, or any action that might affect `wlan1`.
+- End effects are structured notes declared during the turn. They are not arbitrary shell commands.
+
+## Workflow
+
+1. Re-read `/home/arthe/AGENTS.md` and `/home/arthe/workgroup.txt` when coordination may matter.
+2. Run the end step:
+
+```bash
+python3 /home/arthe/.codex/skills/arthexis-cleanup-step/scripts/turn_boundary.py end-step
```
[the remark this answers is not in the document]
Comment (arthexis) [reply skills/arthexis-end-step/SKILL.md]: Addressed in the portability follow-up: end-step skill examples now use home-relative paths instead of hardcoded `/home/arthe` paths. Worker verification also checked no `/home/arthe` matches remain in the owned skill dirs.

**Comment by arthexis** [comment]
Worker follow-up for the fresh Codex review blockers on PR #7421.

Patch pushed: `9c92e32a6` (`Preserve active turn after incomplete cleanup`).

Fixes:
- `PRRT_kwDOPUOpW859lFFd`: `cleanup-step` now keeps `active-turn.json` when live turn-owned processes remain and `--force-kill` was not used. The failed cleanup records the attempt in active state and emits a `cleanup-incomplete` event, so a retry, including `--force-kill`, can still act on the registered process. Successful cleanup and force-kill cleanup still archive and remove active state.
- `PRRT_kwDOPUOpW859lFFg`: `append_event()` now validates the event log path under the turn state root and appends through a directory file descriptor with `O_NOFOLLOW`, so replacing `events.jsonl` with a symlink cannot redirect appends outside the state directory.

Verification run inside Docker container `loomis-pr7421` only:
```bash
python3 -c 'from pathlib import Path; path=Path("skills/arthexis-cleanup-step/scripts/turn_boundary.py"); compile(path.read_text(encoding="utf-8"), str(path), "exec"); print("syntax: ok")'
python3 -m py_compile skills/arthexis-cleanup-step/scripts/turn_boundary.py
HOME=/tmp/loomis-status python3 skills/arthexis-cleanup-step/scripts/turn_boundary.py status
git diff --check
```

Additional targeted probes:
- Synthetic TERM-ignoring process flow: start turn, register PID, run `cleanup-step --timeout 0`, confirm exit `1`, `active-turn.json` still exists, status remains `active`, and cleanup records `still_alive: 1`.
- Retry path: rerun `cleanup-step --timeout 0 --force-kill`, confirm exit `0`, process receives SIGKILL, archive is written, and status returns `no-active-turn`.
- Event-log symlink probe: precreated `events.jsonl -> /tmp/loomis-event-outside.jsonl`, confirmed `start-turn` refuses the path and the outside file is not created.

Residual risk: I did not resolve review threads or mark the PR ready. GitHub checks started after the push and were still running when I checked immediately afterward.

**Comment by arthexis** [comment]
Worker follow-up for the fresh Codex P1 review blockers on PR #7421.

Patch pushed: `80a5c4879` (`Preserve cleanup state for zombies and survivors`).

Fixes:
- `PRRT_kwDOPUOpW859lHJF`: `proc_identity()` now records the `/proc/<pid>/stat` state byte, and live-process matching rejects zombie state `Z` before treating a PID as active turn-owned work. This applies to registered root matching, descendant/live identity collection, and the verified signal path.
- `PRRT_kwDOPUOpW859lHJI`: cleanup now preserves `active-turn.json` whenever `termination["still_alive"]` is non-empty, including `--force-kill` attempts. Incomplete cleanup still records the attempt and emits `cleanup-incomplete`, so retry/status remains possible.

Verification run inside Docker container `harbor-pr7421` only:
```bash
python3 -m py_compile skills/arthexis-cleanup-step/scripts/turn_boundary.py
# synthetic unreaped child zombie probe: process_identity_matches/live_turn_process_* ignore state Z
# helper-level force-kill survivor probe: still_alive after --force-kill keeps active state and writes no archive
# helper-level non-force survivor probe: existing active-state preservation still works
# stale concurrent cleanup probe: changed active turn is not archived over
# event-log symlink probe: append through symlink is refused and outside target is not created
git diff --check
```

Skipped edge case: I used a helper-level forced-survivor probe instead of trying to create a real unkillable container process, because deterministic SIGKILL survivors are not practical in this environment. The probe directly exercises the cleanup branch with `force_kill=True` and a non-empty `still_alive` list.

I did not merge, approve, mark ready, resolve review threads, close PRs/issues, change labels/priority, or add/remove reactions.

**Comment by arthexis** [comment]
Worker follow-up for fresh Codex P1 threads PRRT_kwDOPUOpW859lJB- and PRRT_kwDOPUOpW859lJB_.

Changed:
- Moved cleanup termination behind the post-wait state_lock revalidation, and kept the lock held through signal delivery and state update so a stale cleanup cannot signal processes after the active turn changes.
- Preserved the cleanup boundary for descendants discovered while registered roots are still live by retaining verified process identity snapshots through the wait loop, so a child remains eligible after its registered parent exits.

Verification, all inside Docker container marlin-pr7421:
- python3 -m py_compile skills/arthexis-cleanup-step/scripts/turn_boundary.py
- deterministic ownership-changed-before-termination probe: no termination call and no cross-turn kill
- helper-level registered-root-exits/child-remains probe: descendant tracking continues from retained identity snapshot
- zombie ignored-as-live-work regression probe
- survivor preservation regression probe: active state remains and no archive is written when cleanup still has survivors
- git diff --check

Skipped edge cases: no broader test suite run; this patch is scoped to the owned cleanup script and focused deterministic probes.

**Comment by arthexis** [comment]
Implemented the fresh Codex P1 follow-up for `PRRT_kwDOPUOpW859lLVW`.

Change:
- After `cleanup-step` reacquires `state_lock` for the same turn, it now refreshes live turn-owned process identities from the current active state and unions them into `lingering` before revalidation, termination, archive, or active-state deletion. This covers PIDs registered while cleanup was waiting outside the lock.

Verification, all inside container `rook-pr7421`:
- `python3 -m py_compile skills/arthexis-cleanup-step/scripts/turn_boundary.py`
- Deterministic register-during-wait probe: a PID registered during cleanup wait remains live, cleanup returns incomplete, active state remains present, and no archive is written.
- Survivor preservation regression probe: SIGTERM-ignoring survivor keeps active state and registered process records.
- Stale identity/pre-termination regression probe: stale PID identity snapshots are not signaled.
- `git diff --check`

Skipped edge cases:
- No full suite run; this patch is a one-line cleanup race fix in the standalone skill script, and the focused probes cover the reported race plus the requested regressions.

File: skills/arthexis-cleanup-step/scripts/turn_boundary.py [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+        state["cleanup_step_at"] = now_iso()
+        state["cleanup"] = {
+            "timeout_seconds": timeout_seconds,
+            "live_before": live_before,
+            "lingering_after_wait": sorted(lingering),
+            **termination,
+        }
+        archive_state(state)
+        if ACTIVE_STATE.exists():
+            ACTIVE_STATE.unlink()
```
[the remark this answers is not in the document]
Comment (arthexis) [reply skills/arthexis-cleanup-step/scripts/turn_boundary.py]: Addressed by commit 9c92e32a and retained through the later cleanup fixes: when cleanup still has live processes, active turn state is preserved so retry/status/force cleanup can still operate. Worker verification covered non-force survivor preservation and force-kill retry behavior.

File: skills/arthexis-cleanup-step/scripts/turn_boundary.py [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+            handle.write(content)
+        os.replace(WRITE_TMP_NAME, path.name, src_dir_fd=dir_fd, dst_dir_fd=dir_fd)
+    finally:
+        os.close(dir_fd)
+
+
+def append_event(event: dict[str, Any]) -> None:
+    ensure_state_dir()
+    event = {"time": now_iso(), **event}
+    with EVENT_LOG.open("a", encoding="utf-8") as handle:
```
[the remark this answers is not in the document]
Comment (arthexis) [reply skills/arthexis-cleanup-step/scripts/turn_boundary.py]: Addressed by commit 9c92e32a: event log appends now validate the state root and use a no-follow append path, preserving the cleanup boundary. Worker verification included an event-log symlink refusal probe.

File: skills/arthexis-cleanup-step/scripts/turn_boundary.py [hunk 7]
```diff
@@ -0,0 +1,493 @@
...
+
+
+def process_identity_matches(snapshot: dict[str, Any]) -> bool:
+    pid = snapshot.get("pid")
+    if type(pid) is not int or pid <= 1 or pid == os.getpid():
+        return False
+    identity = proc_identity(pid)
+    if identity is None or identity.get("uid") != os.getuid():
+        return False
+    return identity.get("start_ticks") == snapshot.get("start_ticks")
```
[the remark this answers is not in the document]
Comment (arthexis) [reply skills/arthexis-cleanup-step/scripts/turn_boundary.py:198]: Addressed by commit 80a5c4879: zombie `/proc` entries are no longer treated as live turn-owned processes. Worker verification included an actual unreaped zombie probe plus regression checks.
